adcm_aio_client/core/requesters.py

```diff
@@ -10,32 +10,176 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from typing import AsyncGenerator, Self
+from asyncio import sleep
+from contextlib import suppress
+from dataclasses import dataclass
+from functools import wraps
+from json.decoder import JSONDecodeError
+from typing import Any, Awaitable, Callable, Coroutine, ParamSpec, Self, TypeAlias
+from urllib.parse import urljoin
 
-from typing_extensions import Protocol
+import httpx
 
+from adcm_aio_client.core.errors import (
+    BadRequestError,
+    ConflictError,
+    ForbiddenError,
+    LoginError,
+    NoCredentialsError,
+    NotFoundError,
+    ResponseDataConversionError,
+    ResponseError,
+    RetryRequestError,
+    ServerError,
+    UnauthorizedError,
+    WrongCredentialsError,
+)
+from adcm_aio_client.core.types import Credentials, Requester
 
-class RequesterResponse(Protocol):
-    def as_list(self: Self) -> list: ...
+Json: TypeAlias = Any
+Params = ParamSpec("Params")
+RequestFunc: TypeAlias = Callable[Params, Awaitable["HTTPXRequesterResponse"]]
+DoRequestFunc: TypeAlias = Callable[Params, Awaitable[httpx.Response]]
 
-    def as_dict(self: Self) -> dict: ...
 
+@dataclass(slots=True)
+class HTTPXRequesterResponse:
+    response: httpx.Response
+    _json_data: Json | None = None
 
-class Requester(Protocol):
-    async def get(self: Self, path: str, query_params: dict) -> AsyncGenerator[RequesterResponse]: ...
+    def as_list(self: Self) -> list:
+        if not isinstance(data := self._get_json_data(), list):
+            message = f"Expected a list, got {type(data)}"
+            raise ResponseDataConversionError(message)
 
-    async def post(self: Self, path: str, data: dict) -> AsyncGenerator[RequesterResponse]: ...
+        return data
 
-    async def patch(self: Self, path: str, data: dict) -> AsyncGenerator[RequesterResponse]: ...
+    def as_dict(self: Self) -> dict:
+        if not isinstance(data := self._get_json_data(), dict):
+            message = f"Expected a dict, got {type(data)}"
+            raise ResponseDataConversionError(message)
 
-    async def delete(self: Self, path: str) -> AsyncGenerator[RequesterResponse]: ...
+        return data
 
+    def _get_json_data(self: Self) -> Json:
+        if self._json_data is not None:
```

Comment:
Note for the future: `None` is valid JSON type, so for "unset" cases we may want to use sentinels like https://www.pythonpool.com/sentinel-value-python/ (section "Using Sentinel as a form of Objects
")

but for this case it's not vital, because it'll fail in most cases, because it's not dict or list anyhow

feel free to close this comment on your read